rfc/kubernetes-operator.md

```diff
@@ -0,0 +1,89 @@
+
+We're proposing to move the operator to Spinnaker's GitHub: https://docs.google.com/document/d/1Fs-xK7sOZ4YLSw1ae_cQ_rsB0kMNon9AfYli2a5Udhs/edit?usp=sharing
+
+### Goals and Non-Goals
+
+The goal of the Spinnaker Operator is to enable users to manage Spinnaker via Kubernetes native tools (`kubectl`, ...). Short term we want feature parity with Halyard deployments. Longer term, we'd like extensive preflight checks and active monitoring of Spinnaker.
+
+## Motivation and Rationale
+
+Deploying Spinnaker in Kubernetes with Halyard comes with a few challenges:
+
+- Users often start using Halyard from their laptop and struggle to find a good process to share the configuration.
+- How should Halyard itself be managed? Locally spun up at CI time or as a long-time running service?
+- How do you template the configuration if you have multiple running instances?
+- How do you pass credentials to Halyard, like kubeconfig files or AWS credentials?
+- Users are left with a lot of activities that Halyard doesn’t handle such as exposing Spinnaker and cleaning up old secrets.
+
+Finally, there’s the impression that Spinnaker is not “Kubernetes native” because it cannot be installed and managed with the usual administrator toolchain (kubectl). Talking to Spinnaker administrators, we’ve realized this is a pretty widely shared feeling. They’re already proficient at using kubectl and are reluctant to learn and configure yet another new tool.
+
+## Timeline
+
+Current MVP can deploy and expose Spinnaker. Preflight checks will be added in the coming weeks. Moving the operator to Spinnaker's GitHub account involves publishing new images to gcr and cleaning the build.
+
+## Design
+
+See https://docs.google.com/document/d/1Fs-xK7sOZ4YLSw1ae_cQ_rsB0kMNon9AfYli2a5Udhs/edit#heading=h.7vexvfs9sl5e
+
+### Dependencies
+
+No new dependencies to existing projects. Some (relatively small) changes are necessary in Halyard to:
+
+1) Generate manifests from a dynamic config (not in `~/.hal/`) at runtime
+2) Generate CR and `configMap` from an existing config to facilitate the migration to an operator based deployment.
+
+
+## Drawbacks
+
+It's another project. The operator is also a new deployment running in the user's cluster. If the operator is strictly a tool to configure (vs monitor), that may seem overkill.
+
+## Prior Art and Alternatives
+
+We considered that [operator](https://operatorhub.io/operator/spinnaker-operator) but couldn't find good documentation, the code, or users using it. It also seems to be running a Halyard job in the cluster. Feel free to correct me.
+
+## Known Unknowns
+
+The `spec` of the CRD is still in flux. As we add more features, it will need to evolve before being stable. As an example, new fields are needed to track validation options. The existing `status` is also a first stab.
```

Comment:
Can this link out to the spec of the CRD?  Will the CRD need to be updated as we add config options to the hal config, or does it just reference that as an opaque unit?

Reply:
Added the link (https://github.com/armory/spinnaker-operator/blob/master/deploy/crds/spinnaker_v1alpha1_spinnakerservice_crd.yaml). The CRD does not need to be changed with different versions of Halyard and could potentially work with all recent versions of Halyard (well minus the fact they need to have the manifest endpoint). We just need to edit when we add features to the operator itself (e.g. control how to do validation, how to expose Spinnaker services, etc.)

The Halyard side of things is just a `configMap` (or a `secret` if there's demand) referenced from the `SpinnakerService`.

Reply:
To add to it, it's possible we start pulling some of the fields from the config to the CRD itself to enable a better `kustomize` experience. For instance, `version` is an obvious candidate.